File: backend/src/clawbuddy/workers/ingestion_worker.py
"""Ingestion worker — ARQ background task for document processing.

Replaces: apps/api/src/workers/ingestion.worker.ts (BullMQ)

Pipeline:
1. Download content from MinIO or load inline
2. Split into chunks
3. Embed chunks in batches
4. Store chunks in PostgreSQL + vectors in Qdrant
5. Update document status to READY
"""

from __future__ import annotations

import uuid
from typing import Any

from loguru import logger

from clawbuddy.constants import CHUNK_OVERLAP, CHUNK_SIZE


async def process_document(
    ctx: dict[str, Any],
    document_id: str,
    file_url: str | None = None,
) -> None:
    """ARQ task: process a document through the ingestion pipeline."""
    from clawbuddy.db.models import Document, DocumentChunk
    from clawbuddy.db.session import async_session_factory
    from clawbuddy.lib.sanitize import sanitize_surrogates
    from clawbuddy.services.chunking import chunking_service
    from clawbuddy.services.embedding import embedding_service
    from clawbuddy.services.search import search_service
    from clawbuddy.services.storage import storage_service
    from sqlalchemy import select

    logger.info(f"[Ingestion] Processing document {document_id}")

    async with async_session_factory() as db:
        # Update status to PROCESSING
        result = await db.execute(
            select(Document).where(Document.id == document_id)
        )
        doc = result.scalar_one()
        doc.status = "PROCESSING"
        doc.processing_step = "downloading"
        doc.processing_pct = 0
        await db.commit()

        try:
            # 1. Get text content — from MinIO or inline document content
            if file_url:
                body = await storage_service.download(file_url)
                chunks_raw: list[bytes] = []
                if hasattr(body, "read"):
                    data = await body.read()
                    chunks_raw.append(
                        data if isinstance(data, bytes) else data.encode()
                    )
                elif hasattr(body, "__aiter__"):
                    async for chunk in body:
                        if isinstance(chunk, bytes):
                            chunks_raw.append(chunk)
                        else:
                            chunks_raw.append(chunk.encode())
                else:
                    chunks_raw.append(str(body).encode())
                text = b"".join(chunks_raw).decode("utf-8")
            else:
                # Inline content (e.g. from save_document tool)
                await db.refresh(doc)
                text = doc.content or ""

            # Strip characters that break JSON serialization
            text = sanitize_surrogates(text)

            if not text.strip():
                raise ValueError("Empty document content")

            # 2. Split into chunks
            doc.processing_step = "chunking"
            doc.processing_pct = 15
            await db.commit()

            text_chunks = chunking_service.split_text(
                text, chunk_size=CHUNK_SIZE, overlap=CHUNK_OVERLAP
            )
            logger.info(
                f"[Ingestion] Document {document_id}: {len(text_chunks)} chunks"
            )

            # 3. Ensure Qdrant collection exists
            dimensions = await embedding_service.get_embedding_dimensions()
            await search_service.ensure_collection(dimensions)

            # 4. Generate embeddings and store
            doc.processing_step = "embedding"
            doc.processing_pct = 25
            await db.commit()

            workspace_id = doc.workspace_id
            batch_size = 20
            total_stored = 0

            for i in range(0, len(text_chunks), batch_size):
                batch = text_chunks[i : i + batch_size]
                embeddings = await embedding_service.embed_batch(batch)

                for j, chunk_text in enumerate(batch):
                    qdrant_id = str(uuid.uuid4())
                    chunk_index = i + j

                    safe_content = sanitize_surrogates(chunk_text)

                    # Store in PostgreSQL
                    chunk = DocumentChunk(
                        document_id=document_id,
                        content=safe_content,
                        qdrant_id=qdrant_id,
                        chunk_index=chunk_index,
                        metadata={"workspaceId": workspace_id},
                    )
                    db.add(chunk)
                    await db.flush()

                    # Store vector in Qdrant
                    await search_service.upsert(
                        qdrant_id,
                        embeddings[j],
                        {
                            "documentId": document_id,
                            "chunkId": chunk.id,
                            "chunkIndex": chunk_index,
                            "workspaceId": workspace_id,
                            "content": safe_content[:200],
                        },
                    )

                    total_stored += 1

                # Update progress
                pct = min(25 + int(total_stored / len(text_chunks) * 70), 95)
                doc.processing_step = "indexing"
                doc.processing_pct = pct
                await db.commit()

            # 5. Update document status to READY
            doc.status = "READY"
            doc.content = text[:10000]  # store preview
            doc.chunk_count = total_stored
            doc.processing_step = None
            doc.processing_pct = 100
            await db.commit()

            logger.info(
                f"[Ingestion] Document {document_id} ready: "
                f"{total_stored} chunks indexed"
            )

        except Exception as exc:
            logger.error(
                f"[Ingestion] Failed for document {document_id}: {exc}"
            )
            doc.status = "FAILED"
            doc.processing_step = None
            doc.processing_pct = None
            await db.commit()
            raise

File: apps/api/src/workers/ingestion.worker.ts
import { Worker, type Job } from 'bullmq'
import { redisConnection } from '../lib/redis.js'
import { prisma } from '../lib/prisma.js'
import { storageService } from '../services/storage.service.js'
import { chunkingService } from '../services/chunking.service.js'
import { embeddingService } from '../services/embedding.service.js'
import { searchService } from '../services/search.service.js'
import { CHUNK_SIZE, CHUNK_OVERLAP } from '@clawbuddy/shared'
import { randomUUID } from 'crypto'
import { sanitizeSurrogates } from '../lib/sanitize.js'
import { logger } from '../lib/logger.js'

interface IngestionJobData {
  documentId: string
  fileUrl: string | null
}

const QUEUE_NAME = 'document-ingestion'

const worker = new Worker<IngestionJobData>(
  QUEUE_NAME,
  async (job: Job<IngestionJobData>) => {
    const { documentId, fileUrl } = job.data
    logger.info(`[Ingestion] Processing document ${documentId}`)

    // Update status to PROCESSING
    await prisma.document.update({
      where: { id: documentId },
      data: { status: 'PROCESSING', processingStep: 'downloading', processingPct: 0 },
    })

    try {
      // 1. Get text content — from MinIO or inline document content
      let text: string
      if (fileUrl) {
        const fileStream = await storageService.download(fileUrl)
        const chunksRaw: Buffer[] = []
        for await (const chunk of fileStream as AsyncIterable<string | Uint8Array>) {
          chunksRaw.push(typeof chunk === 'string' ? Buffer.from(chunk) : Buffer.from(chunk))
        }
        text = Buffer.concat(chunksRaw).toString('utf-8')
      } else {
        // Inline content (e.g. from save_document tool)
        const doc = await prisma.document.findUniqueOrThrow({ where: { id: documentId } })
        text = doc.content ?? ''
      }

      // Strip characters that break Prisma JSON serialization
      text = sanitizeSurrogates(text)

      if (!text.trim()) {
        throw new Error('Empty document content')
      }

      // 2. Split into chunks
      await prisma.document.update({
        where: { id: documentId },
        data: { processingStep: 'chunking', processingPct: 15 },
      })

      const chunks = await chunkingService.splitText(text, {
        chunkSize: CHUNK_SIZE,
        overlap: CHUNK_OVERLAP,
      })

      logger.info(`[Ingestion] Document ${documentId}: ${chunks.length} chunks`)

      // 3. Ensure Qdrant collection exists
      const dimensions = await embeddingService.getEmbeddingDimensions()
      await searchService.ensureCollection(dimensions)

      // 4. Generate embeddings and store
      await prisma.document.update({
        where: { id: documentId },
        data: { processingStep: 'embedding', processingPct: 25 },
      })

      const batchSize = 20
      let totalStored = 0

      // Fetch document once for workspaceId (instead of per-chunk)
      const document = await prisma.document.findUniqueOrThrow({ where: { id: documentId } })
      const workspaceId = document.workspaceId

      for (let i = 0; i < chunks.length; i += batchSize) {
        const batch = chunks.slice(i, i + batchSize)
        const embeddings = await embeddingService.embedBatch(batch)

        for (let j = 0; j < batch.length; j++) {
          const qdrantId = randomUUID()
          const chunkIndex = i + j

          // Store in Postgres
          const safeContent = sanitizeSurrogates(batch[j])
          const chunk = await prisma.documentChunk.create({
            data: {
              documentId,
              content: safeContent,
              qdrantId,
              chunkIndex,
              metadata: { workspaceId: workspaceId },
            },
          })

          // Store vector in Qdrant — include chunkId and workspaceId for search
          await searchService.upsert(qdrantId, embeddings[j], {
            documentId,
            chunkId: chunk.id,
            chunkIndex,
            workspaceId: workspaceId,
            content: safeContent.slice(0, 200), // preview
          })

          totalStored++
        }

        // Update progress
        const pct = Math.round(25 + (totalStored / chunks.length) * 70)
        await prisma.document.update({
          where: { id: documentId },
          data: { processingStep: 'indexing', processingPct: Math.min(pct, 95) },
        })
      }

      // 5. Update document status to READY
      await prisma.document.update({
        where: { id: documentId },
        data: {
          status: 'READY',
          content: text.slice(0, 10000), // store preview
          chunkCount: totalStored,
          processingStep: null,
          processingPct: 100,
        },
      })

      logger.info(`[Ingestion] Document ${documentId} ready: ${totalStored} chunks indexed`)
    } catch (error) {
      logger.error(`[Ingestion] Failed for document ${documentId}`, error)
      await prisma.document.update({
        where: { id: documentId },
        data: { status: 'FAILED', processingStep: null, processingPct: null },
      })
      throw error
    }
  },
  {
    connection: redisConnection,
    concurrency: 3,
  },
)

worker.on('completed', (job) => {
  logger.info(`[Ingestion] Job ${job.id} completed`)
})

worker.on('failed', (job, err) => {
  logger.error(`[Ingestion] Job ${job?.id} failed`, err)
})

export { worker as ingestionWorker }
